crates/rspack_plugin_javascript/src/visitors/mod.rs

```diff
@@ -55,108 +198,44 @@ pub fn run_before_pass(
-        ),
-        should_transform_by_react
-      ),
-      Optional::new(
-        rspack_swc_visitors::fold_react_refresh(unresolved_mark),
-        should_transform_by_react
-          && options
-            .builtins
-            .react
-            .refresh
-            .and_then(|v| if v { Some(v) } else { None })
-            .is_some()
-      ),
-      either!(
-        options.builtins.emotion,
-        |emotion_options: &rspack_swc_visitors::EmotionOptions| {
-          rspack_swc_visitors::emotion(
-            emotion_options.clone(),
-            &resource_data.resource_path,
-            xxh32(source.as_bytes(), 0),
-            cm.clone(),
-            comments,
-          )
-        }
-      ),
-      either!(options.builtins.relay, |relay_option| {
-        rspack_swc_visitors::relay(
-          relay_option,
-          resource_data.resource_path.as_path(),
-          PathBuf::from(AsRef::<Path>::as_ref(&options.context)),
-          unresolved_mark,
-        )
-      }),
-      either!(options.builtins.plugin_import, |options| {
-        rspack_swc_visitors::import(options)
-      }),
-      // enable if configurable
-      // swc_visitor::const_modules(cm, globals),
-      Optional::new(
-        rspack_swc_visitors::define(&options.builtins.define),
-        !options.builtins.define.is_empty()
-      ),
-      Optional::new(
-        rspack_swc_visitors::provide_builtin(&options.builtins.provide, unresolved_mark),
-        !options.builtins.provide.is_empty()
-      ),
-      Optional::new(
-        swc_visitor::export_default_from(),
-        syntax.export_default_from()
+      builtins_additional_feature_transforms(
+        resource_data,
+        options,
+        module_type,
+        source,
+        top_level_mark,
+        unresolved_mark,
+        cm
       ),
-      // enable if necessary
-      // swc_visitor::simplifier(unresolved_mark, Default::default()),
-      // enable if configurable
-      // swc_visitor::json_parse(min_cost),
-      swc_visitor::paren_remover(comments.map(|v| v as &dyn Comments)),
-      // es_version
       Optional::new(
-        swc_visitor::compat(
-          options.builtins.preset_env.clone(),
-          es_version,
-          assumptions,
+        compat_transform(
+          resource_data,
+          options,
           top_level_mark,
           unresolved_mark,
-          comments,
-          syntax.typescript()
+          assumptions,
+          syntax
         ),
-        !resource_path.contains("@swc/helpers")
-          && !resource_path.contains("tslib")
-          && !resource_path.contains("core-js")
+        options.should_transform_by_default()
       ),
-      swc_visitor::reserved_words(),
+      builtins_webpack_plugin(options, unresolved_mark),
+      // This will be deprecated in the future when TypeScript support and transform by default is dropped.
+      // These features are coupled with this.
       swc_visitor::inject_helpers(unresolved_mark),
```

Comment:
`inject_helpers` are working with `compat_transform` and `decorator` so this cannot be removed now.